org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTParser.java

```diff
@@ -1229,68 +1233,28 @@ private ASTNode internalCreateASTCached(IProgressMonitor monitor) {
 						throw new IllegalStateException();
 					}
 					if ((this.bits & CompilationUnitResolver.PARTIAL) != 0) {
-						searcher = new NodeSearcher(this.focalPointPosition);
+						useSearcher = true;
 					}
 					int flags = 0;
 					if ((this.bits & CompilationUnitResolver.STATEMENT_RECOVERY) != 0) {
 						flags |= ICompilationUnit.ENABLE_STATEMENTS_RECOVERY;
 					}
-					if (searcher == null && ((this.bits & CompilationUnitResolver.IGNORE_METHOD_BODIES) != 0)) {
+					if (!useSearcher && ((this.bits & CompilationUnitResolver.IGNORE_METHOD_BODIES) != 0)) {
 						flags |= ICompilationUnit.IGNORE_METHOD_BODIES;
 					}
+
 					if (needToResolveBindings) {
 						if ((this.bits & CompilationUnitResolver.BINDING_RECOVERY) != 0) {
 							flags |= ICompilationUnit.ENABLE_BINDINGS_RECOVERY;
 						}
-						try {
-							// parse and resolve
-							compilationUnitDeclaration =
-								CompilationUnitResolver.resolve(
-									sourceUnit,
-									this.project,
-									getClasspath(),
-									searcher,
-									this.compilerOptions,
-									this.workingCopyOwner,
-									flags,
-									monitor);
-						} catch (JavaModelException e) {
-							flags &= ~ICompilationUnit.ENABLE_BINDINGS_RECOVERY;
-							compilationUnitDeclaration = CompilationUnitResolver.parse(
-									sourceUnit,
-									searcher,
-									this.compilerOptions,
-									flags);
-							needToResolveBindings = false;
-						}
-					} else {
-						compilationUnitDeclaration = CompilationUnitResolver.parse(
-								sourceUnit,
-								searcher,
-								this.compilerOptions,
-								flags,
-								this.project);
-						needToResolveBindings = false;
 					}
-					CompilationUnit result = CompilationUnitResolver.convert(
-						compilationUnitDeclaration,
-						sourceUnit.getContents(),
-						this.apiLevel,
-						this.compilerOptions,
-						needToResolveBindings,
-						wcOwner,
-						needToResolveBindings ? new DefaultBindingResolver.BindingTables() : null,
-						flags,
-						monitor,
-						this.project != null,
-						this.project);
-					result.setTypeRoot(this.typeRoot);
-					return result;
+
+					CompilationUnit result2 = this.unitResolver.toCompilationUnit(sourceUnit, needToResolveBindings, this.project, getClasspath(), useSearcher ? this.focalPointPosition : -1, this.apiLevel, this.compilerOptions, this.workingCopyOwner, wcOwner, flags, monitor);
```

Comment:
What is the reason for extracting this part of the logic into a new API `toCompilationUnit`?

Reply:
I guess it's more convenience. What do you think would be better here? Making the whole `internalCreateASTForKind` abstract or part of the interface?

Reply:
The CompilationUnitResolver provides three basic operations: parse, resolve, and convert. I'm wondering why we don't simply extract these operations into the new API. Then, we would only need to replace the resolver instance in the ASTParser.

Reply:
The issue with convert is that it is parser specific. So we cannot easily `convert` as it in a parser-independent interface. I think that's why we extract `toCompilationUnit` in 1st place, to hide the internal parser objects and the parser-specific conversion work.

Reply:
The `convert` method has `DefaultBindingResolver.BindingTables bindingTables` as one of its parameters. Not only is this internal, but we also intend on using a binding resolver class, and so forcing this parameter into the interface is inappropriate. 

I believe the `toCompilationUnit` method is appropriate as designed here. It has a signature that allows extenders the most flexibility with no extraneous parameters that wouldn't make sense for a custom implementation.

Reply:
Understood, the new API `toCompilationUnit` makes sense. A side note is the new public API uses two internal ECJ classes such as `org.eclipse.jdt.internal.compiler.env.ICompilationUnit` and `org.eclipse.jdt.internal.compiler.batch.FileSystem.Classpath`. How about passing the raw data (contents, ... sourcepaths, ...) or wrapping them in records?

For example, 
`public record SourceUnit(contents, packageName, fileName, javaElement)`
`public record Classpath(sourcepaths, classpaths, includeVMBootpath)`

Reply:
I'm not sure it would be immediately profitable. Those internal compiler classes are likely to stay for a while in practice and it could be a bit too early to refactor this. Maybe later we'll have some concrete case as to why *not* requiring them, but we're not there yet IMO, so it's simpler and faster to stick to those (useful) classes for an internal/experimental extension point which is constrained on many aspects already.

As to using a record, then one issue would be that we would have to actually resolve all values at creation. I'm not so familiar with implementation of eg ICompilationUnit, but I'm wondering what if those are dynamic; more dynamic than what a record can capture? This is also 1 reason why I think refactoring that immediately may not be best.

IMO, it would be better to keep the current state of this proposal, make annotations here and there wherever it makes sense to clearify that all that is experimental/provisional; and then let more experience lead to improving the design later.

Reply:
Unit tests have been added